oozie-to-airflow/mappers/fs_mapper.py

```diff
@@ -0,0 +1,198 @@
+
+    return command
+
+
+def prepare_move_command(node: Element, params):
+    source = normalize_path(node.attrib[FS_TAG_SOURCE], params)
+    target = normalize_path(node.attrib[FS_TAG_TARGET], params)
+
+    command = "fs -mv {source} {target}".format(source=source, target=target)
+    return command
+
+
+def prepare_chmod_command(node: Element, params):
+    path = normalize_path(node.attrib[FS_TAG_PATH], params)
+    permission = node.attrib[FS_TAG_PERMISSIONS]
+    # TODO: Add support for dirFiles
+    # dirFiles = bool_value(node, FS_TAG_DIRFILES)
+    recursive = node.find(FS_TAG_RECURSIVE) is not None
+    extra_param = "-R" if recursive else ""
+
+    command = "fs -chmod {extra} {permission} {path}".format(
+        extra=extra_param, path=shlex.quote(path), permission=shlex.quote(permission)
+    )
+    return command
+
+
+def prepare_touchz_command(node: Element, params):
+    path = normalize_path(node.attrib[FS_TAG_PATH], params)
+
+    command = "fs -touchz {path}".format(path=shlex.quote(path))
+    return command
+
+
+def prepare_chgrp_command(node: Element, params):
+    path = normalize_path(node.attrib[FS_TAG_PATH], params)
+    group = normalize_path(node.attrib[FS_TAG_GROUP], params)
+
+    recursive = node.find(FS_TAG_RECURSIVE) is not None
+    extra_param = "-R" if recursive else ""
+
+    command = "fs -chgrp {extra} {group} {path}".format(
+        extra=extra_param, group=shlex.quote(group), path=shlex.quote(path)
+    )
+    return command
+
+
+def prepare_setrep_command(node: Element, params):
+    path = normalize_path(node.attrib[FS_TAG_PATH], params)
+    fac = node.attrib[FS_TAG_REPLFAC]
+
+    command = "fs -setrep {fac} {path}".format(fac=shlex.quote(fac), path=shlex.quote(path))
+    return command
+
+
+FS_OPERATION_MAPPERS = {
+    FS_OP_MKDIR: prepare_mkdir_command,
+    FS_OP_DELETE: prepare_delete_command,
+    FS_OP_MOVE: prepare_move_command,
+    FS_OP_CHMOD: prepare_chmod_command,
+    FS_OP_TOUCHZ: prepare_touchz_command,
+    FS_OP_CHGRP: prepare_chgrp_command,
+    FS_OP_SETREP: prepare_setrep_command,
+}
+
+
+class Task(NamedTuple):
+    task_id: str
+    rendered_template: str
+
+
+def chain(ops):
+    return [Relation(from_task_id=a.task_id, to_task_id=b.task_id) for a, b in zip(ops, ops[1::])]
+
+
+class FsMapper(ActionMapper):
+    """
+    Converts a FS Oozie node to an Airflow task.
+    """
+
+    tasks: List[Task]
+
+    def __init__(self, *args, **kwargs):
+        super().__init__(*args, **kwargs)
+        self.tasks = []
+
+    def on_parse_node(self):
+        super().on_parse_node()
+        self.tasks = self.parse_tasks()
+
+    def parse_tasks(self):
+        if not list(self.oozie_node):
+            return [
+                Task(
+                    task_id=self.name,
+                    rendered_template=render_template(
+                        "dummy.tpl", task_id=self.name, trigger_rule=self.trigger_rule
```

Comment:
Why are we rendering a dummy operator here? Shouldn't we skip rendering it at all in such a case?

Reply:
This is required for relations. Other tasks can have relations to this node.  In future, we can optimize this case, but for now it is good enough to create an empty task.  Such optimization would have to change the relations of other nodes.

It is worth noting that there is a small chance that the real workflow will contain empty ooze nodes

Reply:
I created a diagram that illustrates the problem.
![Untitled Diagram](https://user-images.githubusercontent.com/12058428/56145004-c4f1ce00-5fa3-11e9-8d42-a1a7199e3a6f.png)
Do you know a simple solution to this problem?

Reply:
So this is the "FS Mapper without tasks" case that we're solving here?

Reply:
Yes. You're right.

Reply:
Ok, thanks for the explanation and for the diagram :+1: